Avoid crash in wxBitmap::GetImage() for invalid bitmaps

vadz opened the pull request:
Just return NULL from this (wxOSX private, in spite of not using a
port-specific prefix) method.

This fixes crash in wxButton::SetBitmapXXX(wxNullBitmap), as shown by
the new test case which used to crash but doesn't do it any longer.

Closes [#19257](https://trac.wxwidgets.org/ticket/19257).

---

Stefan, I think it's more natural and safer to fix it like this rather than avoiding calls to `GetNSImage()` for empty bitmaps, as there are quite a few calls to it in the code and it seems that at least a couple of others could result in a crash as well. If you don't mind this change, I'm also going to go through the other ones and maybe remove the redundant checks for `bitmap.IsOk()`.

Alternatively, if you do mind, I can replace the check here with an assert (`wxCHECK`) and add check to `wxButton` code too.

vadz committed:
Avoid crash in wxBitmap::GetImage() for invalid bitmaps

Just return NULL from this (wxOSX private, in spite of not using a
port-specific prefix) method.

This fixes crash in wxButton::SetBitmapXXX(wxNullBitmap), as shown by
the new test case which used to crash but doesn't do it any longer.

Closes #19257.

csomor reviewed (approved):
Hi, I'm absolutely fine with this, thanks a lot.

vadz committed:
Remove always true check from assert in MSW wxAnyButton code

The bitmap here is always valid as the invalid case is handled in the
beginning of the function.

vadz committed:
Return empty bitmaps from wxButton if not explicitly set in wxMSW

Do this for consistency with the other ports and because this seems more
useful anyhow.

Update the documentation to make this behaviour more clear and document
this change as a (minor) incompatibility in wxMSW.

Also add more unit tests to check for this behaviour. Note this also
fixes the problem with the unit test added in the grandparent commit
under MSW.

vadz committed:
Use wxBitmap::GetUIImage() in wxiOS code

It looks like this should be the right function to use, rather than the
more generic GetImage() which is going to be renamed in the upcoming
commits.

vadz committed:
Use wxBitmap::GetNSImage() in Cocoa code

Use Cocoa-specific function rather than generically-named GetImage().

No real changes.

vadz committed:
Rename wxBitmap::GetImage() to OSXGetImage()

Make it more clear that this function is specific to Mac ports.

vadz committed:
Remove unnecessary checks before calling GetNSImage()

Now that this function is guaranteed to return NULL if the bitmap is
invalid, the code can be simplified to rely on it.

No real changes.

vadz commented:
Thanks, I've done some more cleanup and also fixed wxMSW to behave as wxGTK and wxOSX for the button bitmaps.

There are some mildly incompatible changes here, notably `wxBitmap::GetImage()` has been renamed to `OSXGetImage()` in wxOSX, but I think it should be fine and I'd like to do it before 3.2 because I don't like having a Mac-only function without this prefix.

If nobody objects, I'll merge this soon.

vadz commented:
P.S. CI failure is unrelated, but there is a real bug in the Unix version of `wxWebRequest` that would be really nice to run under TSAN...

csomor commented:
> There are some mildly incompatible changes here, notably `wxBitmap::GetImage()` has been renamed to `OSXGetImage()` in wxOSX, but I think it should be fine and I'd like to do it before 3.2 because I don't like having a Mac-only function without this prefix.
> 
> If nobody objects, I'll merge this soon.

I thought using 'generic' types as WXImage would qualify as a generic API, because also other ports could implement it the same way, if wanted/needed. But if you want it to be clear from the name that it currently only exists on OSX that's fine for me as well.

vadz commented:
> I thought using 'generic' types as `WXImage` would qualify as a generic API, because also other ports could implement it the same way, if wanted/needed.

Thanks for the explanation, but we have `GetHandle()` in wxMSW for things like this and I just dislike using `Image` in the method name because IMO it's confusing due to the overlap with `wxImage`, so I'd like to get rid of this one.
